client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.elasticsearch.client.asyncsearch;
+
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchType;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.client.Validatable;
+import org.elasticsearch.client.ValidationException;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A request to track asynchronously the progress of a search against one or more indices.
+ */
+public class SubmitAsyncSearchRequest implements Validatable {
+
+    public static long MIN_KEEP_ALIVE = TimeValue.timeValueMinutes(1).millis();
+
+    private TimeValue waitForCompletion;
+    private Boolean cleanOnCompletion;
+    private TimeValue keepAlive;
+    private final SearchRequest searchRequest;
+
+    /**
+     * Creates a new request
+     */
+    public SubmitAsyncSearchRequest(SearchSourceBuilder source, String... indices) {
+        this.searchRequest = new SearchRequest(indices, source);
+        searchRequest.setCcsMinimizeRoundtrips(false);
+        searchRequest.setPreFilterShardSize(1);
+        searchRequest.setBatchedReduceSize(5);
+        searchRequest.requestCache(true);
+    }
+
+
+    /**
+     * Get the {@link SearchRequest} that this submit request wraps
+     */
+    public SearchRequest getSearchRequest() {
```

Comment:
I see that whether or not to expose the inner search request was discussed in previous reviews. I am confused though on what direction was chosen. I see that the getters and setters from the search request are copied to the async request, but the inner search request is still exposed through this getter and can be modified directly. Didn't we want to rather hide it from users?

Reply:
+1 to hide

Reply:
Okay, I will look into that. Will make testing a bit more awkward though because I can't reuse some existing infra then.

Reply:
package protected ? ;)

Reply:
Work work but then I need to move some classes around. I'd prefer that to adding all those getters to the submit request.

Reply:
I think that if we can do without the getter, even better. The changes you had to make to remove it don't look so bad. The problem with keeping it package protected is that theoretically some stuff can still be set directly to the search request, then we need to validate parameters that are not even supported.

Reply:
This is proving to be tricky. I cannot easily move the AsyncSearchRequestConverters out of the `org.elasticsearch.client` package since they need package private infra from RequestConverters. I can probably move the `SubmitAsyncSearchRequest` into that package though, will take a look what that would change.

Reply:
I pushed f686a50 which moves the new request and response classes into `org.elasticsearch.client` to be able to use a package protected `getSearchRequest()` to avoid all the boilerplate getters on the new request. I'm unsure what I like best, take a look and let me know which direction you are leaning.

Reply:
I  think no getter is  safer. We also do have extra protection on the server side for unsupported values but I think the client should never allow to send unsupported values.

Reply:
I agree with Luca, we should avoid the getter on the search request. However, we should have a getter for all options that we expose so I don't see why you want to avoid them ? If we have a setter, we need to provide a way to access the value.

Reply:
Ok, I reverted that change back to the one where I removed `getSearchRequest` and added a bunch of getters.